[#64550] Update ng `attribute-help-text` to use Primerized attribute help text dialog

myabc opened the pull request:
⚠️ **~~This PR is based off #19091. Please review/merge that PR first.~~**

# Ticket

https://community.openproject.org/wp/64550

# What are you trying to accomplish?

Updates ng `attribute-help-text` component to invoke Primerized dialogs, harmonising the look and feel of attribute help texts.

This work also includes a couple global fixes to prevent multiple dialogs from being created in the help text button is clicked multiple times in quick succession.

## Screenshots

![Screenshot 2025-06-26 at 13 16 08](https://github.com/user-attachments/assets/f7dc85dd-830f-419b-a353-15f2959c4f53)

# What approach did you choose and why?

Approaches to prevent multiple dialogs from being created:

1. **at the dialog creation level** (unique dialogs): The Turbo Stream dialog action has been updated to prevent multiple `<dialog>` elements with the same `id` from being appended to `document.body`. If a dialog with a  given id is already present in the DOM then the content/`innerHTNL` of the dialog is updated instead (via Turbo morphing). Also worth noting:

    - Event listeners are not rebound. Existing handlers on the dialog (e.g. `close`) are preserved.
    - `dialog.showModal()` is always called by the dialog action (it should be idempotent) to handle the case where the dialog is closed but still present in the DOM _(why would this happen?)_.


2. **at the HTTP request level** (debouncing): `TurboRequestsService#request` now supports a `requestId` parameter. Passing a `requestId` will abort any previous requests with the same id. Additionally, the higher-level convenience methods `#submitForm` and `#requestStream` have been updated to automatically abort any previous requests with the same URL (and search params). **This is based on the assumption that there should only be one request in flight per URL.**

4. **at the component level**: the button is disabled on click and only re-enabled once the Promise returned by the modal service `show` has settled. _(check the styling? do we need to make it consistent with Turbo links?)_

### Potential further improvements

Extract parts of `AttributeHelpTextModalService` (and possibly `JobStatusModalService`) into a generic and reusable `PrimerDialogService` or `AsyncDialogService`.

# Merge checklist

- [x] Upstream changes to `ToolTipElement` - primer/view_components#3558
  - [x] PR created
  - [x] PR merged
  - [x] Primer released
- [x] Added/updated tests
- [ ] Added/updated documentation in Lookbook (patterns, previews, etc)
- [ ] Tested major browsers (Chrome, Firefox, Edge, ...)

myabc committed:
[#64550] Port attribute-help-text to Primer dialog

Makes ng `attribute-help-text` interop with Primerized attribute help
text dialogs.

myabc committed:
Remove unused Components::AttributeHelpTextModal

HDinger reviewed (commented):
I managed quite easy to open the dialog twice. When I first loaded the full view it took really long to open the dialog so I clicked again and again on the question mark. I ended up with three dialogs above each other. After that initial load, the dialog was faster but I still could open two dialogs with a double click. Maybe we can cancel one of the events in case of a double click?

myabc committed:
Add initial AttributeHelpTextModalService spec

Currently tests the happy path only.

myabc commented:
@HDinger as mentioned on Element, I wasn't able to reproduce opening multiple instances of the dialog via double-clicking. Perhaps we can take a look together? In the meantime, I'm putting up for another review.

I've added some additional specs to ensure parity with the ViewComponent implementation, as well as tests for invoking Primer dialogs from Angular.

HDinger reviewed (commented):
The changes look good 👍 

Regarding the doubled dialog. It mostly only happens after navigating to a WP. I assume that there is still some data for the WP view loaded in the background making the whole page a tad slower which in the end results in the two events being fired. It is however not happening every time. I recorded a screen cast. Note, that on the first try it works correclty and only on the second try, two dialogs appear.
![Jul-02-2025 08-42-00](https://github.com/user-attachments/assets/4af8b43d-a2b7-404b-bbf7-95c76eebf4ab)

myabc committed:
Fix creation of multiple dialogs with same id

Updates Dialog Stream action to append a new dialog (`<dialog-helper>`
and `<dialog>` elements) only when there is no existing dialog with that
id in the DOM.

If there is an existing dialog, the dialog contents are updated instead
(via morphing).

myabc committed:
Teach TurboRequestsService how to abort requests

Requests can be aborted individually by passing a `requestId`.

myabc committed:
Make submitForm, requestStream auto abort requests

Automatically aborts previous requests by defaulting `requestId` to the
request URL. This is based on an assumption that there should only be
one request in flight for a given request URL.

myabc committed:
Apply unique IDs for each help text dialog

(Theoretically) permits multiple help text dialogs in DOM (but only one
per help attribute help text modal instance).

myabc committed:
Make requestStream show Turbo progress bar

myabc committed:
Disable help text button on click

Prevents multiple calls to service/HTTP requests on successive clicks.

myabc committed:
Flesh out AttributeHelpTextModalService spec

myabc commented:
@HDinger as we discussed yesterday, there are a couple different ways of mitigating multiple dialogs being created by multiple, successive user clicks.

I've gone ahead and implemented three different solutions and detailed them above in the description under Approach.

I realise that that might seem a bit like overkill, but two mitigations are at the global level and I would consider them "good practice" rather than defensive programming. Still, it's probably a good idea to get a few more eyes on this PR to ensure that these changes don't have unintentional side-effects for other components/functionality.

Additionally, since you last reviewed, I've added more specs for both the component and modal service. I've also enabled the Turbo progress bar when making requests via `requestStream`.

HDinger reviewed (approved):
Lgtm 👍 Thanks for investing the time to fix the double rendering issue 🙇
